### PR TITLE
Ps 5.7 tdb 129

### PR DESCRIPTION
**Pull request by george-lorch**
TDB-134 : TokuDB 5.7 is missing minor optimization for read-only transactions.
TDB-135 : TokuDB 5.7 missing mrr implementation as it seems compiled out of 5.7
TDB-129 : Normalize TokuDB ALTER code to remove 5.5 and version dependent compilation
TDB-129 : 'Fix' tokub conditional compilation dependent on version id starting in 5.7

### CODE REVIEW

**Comment by george-lorch** [comment]
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1725/

File: storage/tokudb/ha_tokudb_alter.cc [hunk 1]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
 
 #ident "Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved."
 
-#if TOKU_INCLUDE_ALTER_56
-
-#if 100000 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 100099
-#define TOKU_ALTER_RENAME ALTER_RENAME
-#elif (50600 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 50699) || \
-      (50700 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 50799)
 #define TOKU_ALTER_RENAME ALTER_RENAME
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Remove as well?

File: storage/tokudb/ha_tokudb_alter.cc [hunk 2]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
-#error
-#endif
 
-#include "ha_tokudb_alter_common.cc"
 #include <sql_array.h>
 #include <sql_base.h>
 
 #include <vector>
 
+TOKUDB_UNUSED(static bool tables_have_same_keys(TABLE* table,
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Remove TOKUDB_UNUSED, it's actually used?

File: storage/tokudb/ha_tokudb_alter.cc [hunk 3]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+    retval = true;
+cleanup:
+    return retval;
+}
+
+// MySQL sets the null_bit as a number that you can bit-wise AND a byte to
+// to evaluate whether a field is NULL or not. This value is a power of 2, from
+// 2^0 to 2^7. We return the position of the bit within the byte, which is
+// lg null_bit
+TOKUDB_UNUSED(static inline uint32_t get_null_bit_position(uint32_t null_bit));
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 4]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+            retval = 7;
+            break;
+        default:
+            assert_unreachable();
+    }
+    return retval;
+}
+
+// returns the index of the null bit of field.
+TOKUDB_UNUSED(
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 5]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+    static inline uint32_t get_overall_null_bit_position(TABLE* table,
+                                                         Field* field));
+static inline uint32_t get_overall_null_bit_position(TABLE* table,
+                                                     Field* field) {
+    uint32_t offset = get_null_offset(table, field);
+    uint32_t null_bit = field->null_bit;
+    return offset * 8 + get_null_bit_position(null_bit);
+}
+
+// not static since 51 uses this and 56 does not
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Outdated comment?

File: storage/tokudb/ha_tokudb_alter.cc [hunk 6]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+TOKUDB_UNUSED(static bool are_null_bits_in_order(TABLE* table));
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Remove TOKUDB_UNUSED ?

File: storage/tokudb/ha_tokudb_alter.cc [hunk 7]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+                break;
+            }
+            first = false;
+            curr_null_pos = pos;
+        }
+    }
+    return retval;
+}
+
+TOKUDB_UNUSED(static uint32_t get_first_null_bit_pos(TABLE* table));
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 8]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+        bool nullable = (curr_field->null_bit != 0);
+        if (nullable) {
+            table_pos = get_overall_null_bit_position(table, curr_field);
+            break;
+        }
+    }
+    return table_pos;
+}
+
+TOKUDB_UNUSED(static bool is_column_default_null(TABLE* src_table,
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 9]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+    for (uint i = 0; i < table_a->s->fields; i++) {
+        Field* field_a = table_a->field[i];
+        Field* field_b = table_b->field[i];
+        if (!fields_are_same_type(field_a, field_b))
+            return false;
+    }
+    return true;
+}
+
+TOKUDB_UNUSED(static bool column_rename_supported(TABLE* orig_table,
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 10]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+    if (!keys_same_for_cr) {
+        retval = false;
+        goto cleanup;
+    }
+    retval = true;
+cleanup:
+    return retval;
+}
+
+TOKUDB_UNUSED(static int find_changed_columns(uint32_t* changed_columns,
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 11]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+        changed_columns[curr_num_changed_columns] = i;
+        curr_num_changed_columns++;
+    }
+    *num_changed_columns = curr_num_changed_columns;
+    retval = 0;
+cleanup:
+    return retval;
+}
+
+TOKUDB_UNUSED(static bool tables_have_same_keys_and_columns(TABLE* first_table,
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Likewise

File: storage/tokudb/ha_tokudb_alter.cc [hunk 12]
```diff
@@ -23,25 +23,713 @@ Copyright (c) 2006, 2015, Percona and/or its affiliates. All rights reserved.
...
+        retval = false;
+        goto exit;
+    }
+
+    retval = true;
+exit:
+    return retval;
+}
+
+#if TOKU_INCLUDE_WRITE_FRM_DATA
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc]: Remove this block?

File: storage/tokudb/ha_tokudb_alter.cc [hunk 13]
```diff
@@ -898,22 +1578,16 @@ bool ha_tokudb::commit_inplace_alter_table(
...
 
     if (commit) {
-#if (50613 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 50699) || \
-    (50700 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 50799) || \
-    (100000 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 100099)
         if (ha_alter_info->group_commit_ctx) {
             ha_alter_info->group_commit_ctx = NULL;
         }
-#endif
-#if (50500 <= MYSQL_VERSION_ID && MYSQL_VERSION_ID <= 50599) || \
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb_alter.cc:915]: Inside of this block should have been removed too?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: storage/tokudb/ha_tokudb_alter.cc (hunk 1 above)
Comment (george-lorch) [reply storage/tokudb/ha_tokudb_alter.cc]: Oops, yes. Not sure how this slipped through grep foo

File: storage/tokudb/ha_tokudb_alter.cc (hunk 2 above)
Comment (george-lorch) [reply storage/tokudb/ha_tokudb_alter.cc]: Ahh, maybe. I never considered these. Will have to validate each one.

File: storage/tokudb/ha_tokudb_alter.cc (hunk 13 above)
Comment (george-lorch) [reply storage/tokudb/ha_tokudb_alter.cc:915]: That is a good catch. This would have introduced a real regression.

**Comment by george-lorch** [comment]
OK, fixed up the missed changes as well as the potential regression. New jenkins here https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1728/

File: storage/tokudb/ha_tokudb.cc [hunk 14]
```diff
@@ -478,14 +478,7 @@ static inline bool do_ignore_flag_optimization(
 
 #if TOKU_INCLUDE_EXTENDED_KEYS
 static inline uint get_ext_key_parts(const KEY *key) {
```
Comment (laurynas-biveinis) [inline storage/tokudb/ha_tokudb.cc]: Consider inlining (similar to close_dsmrr() -> ds_mrr.dsmrr_close() change)

**Review by laurynas-biveinis — APPROVED** [review]
A minor comment; push with or without addressing it

File: storage/tokudb/ha_tokudb.cc (hunk 14 above)
Comment (george-lorch) [reply storage/tokudb/ha_tokudb.cc]: Yeah, that makes sense, it is only used in one place so no real need for stand alone function.

**Comment by peterschwaller** [comment]
touch

**Pull request merged**